src/algorithms/calorimetry/CalorimeterHitsMerger.cc

```diff
@@ -41,24 +43,57 @@ void CalorimeterHitsMerger::init() {
         return;
     }
 
+    // initialize descriptor + decoders
     try {
-        auto id_desc = m_detector->readout(m_cfg.readout).idSpec();
-        id_mask = 0;
-        std::vector<std::pair<std::string, int>> ref_fields;
-        for (size_t i = 0; i < m_cfg.fields.size(); ++i) {
-            id_mask |= id_desc.field(m_cfg.fields[i])->mask();
-            // use the provided id number to find ref cell, or use 0
-            int ref = i < m_cfg.refs.size() ? m_cfg.refs[i] : 0;
-            ref_fields.emplace_back(m_cfg.fields[i], ref);
-        }
-        ref_mask = id_desc.encode(ref_fields);
+      id_desc = m_detector->readout(m_cfg.readout).idSpec();
+      id_decoder = id_desc.decoder();
+      for (const auto& field : m_cfg.fields) {
+        const short index = id_decoder->index(field);
+      }
     } catch (...) {
-        auto mess = fmt::format("Failed to load ID decoder for {}", m_cfg.readout);
-        warning(mess);
+      auto mess = fmt::format("Failed to load ID decoder for {}", m_cfg.readout);
+      warning(mess);
 //        throw std::runtime_error(mess);
     }
-    id_mask = ~id_mask;
-    debug("ID mask in {:s}: {:#064b}", m_cfg.readout, id_mask);
+
+    // if no field-by-field mappings provided, initialize relevant bitmasks
+    // otherwise intialize relevant functionals
+    if (m_cfg.mappings.empty()) {
+      id_mask = 0;
+      std::vector<RefField> ref_fields;
+      for (size_t i = 0; i < m_cfg.fields.size(); ++i) {
+          id_mask |= id_desc.field(m_cfg.fields[i])->mask();
+          // use the provided id number to find ref cell, or use 0
+          int ref = i < m_cfg.refs.size() ? m_cfg.refs[i] : 0;
+          ref_fields.emplace_back(m_cfg.fields[i], ref);
+      }
+      ref_mask = id_desc.encode(ref_fields);
+      id_mask = ~id_mask;
+      debug("ID mask in {:s}: {:#064b}", m_cfg.readout, id_mask);
+    } else {
+
+      // lambda to translate IDDescriptor fields into function parameters
+      std::function hit_to_map = [this](const edm4eic::CalorimeterHit& hit) {
+        std::unordered_map<std::string, double> params;
+        for (const auto& name_field : id_desc.fields()) {
+          params.emplace(name_field.first, name_field.second->value(hit.getCellID()));
+          trace("{} = {}", name_field.first, name_field.second->value(hit.getCellID()));
+        }
+        return params;
+      };
+
+      // intialize functions
+      //   - NOTE this assumes provided fields are 1-to-1!
+      auto& svc = algorithms::ServiceSvc::instance();
+      for (std::size_t iMap = 0; const auto& mapping : m_cfg.mappings) {
+        if (iMap < m_cfg.fields.size()) {
```

Comment:
This also breaks my brain a bit. A check for matching length and a 1-to-1 loop would be nicer.

Reply:
Sure! This was trying to work in the above design choice that providing a map makes the algorithm ignores any provided bitmasks.